### PR TITLE
Check type_index of runtime visible type annotations

### PR DESCRIPTION
**Pull request by pdbain-ibm**
If a type annotation's type_index does not refer to a CONSTANT_UTF8_info
structure, don't mark it as being used.  When writing the ROM class, ensure
that constant pool entries marked as used by annotations are UTF8.

FYI @fengxue-IS 

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

**Comment by pdbain-ibm** [comment]
Please see https://github.com/eclipse/openj9/pull/1579 for comments on the issue and previous solution.

**Comment by pdbain-ibm** [comment]
@fengxue-IS @DanHeidinga @gacholio would you kindly review this?

**Comment by pdbain-ibm** [comment]
@pshipton would you please launch PR builds?

File: runtime/bcutil/ROMClassWriter.cpp [hunk 1]
```diff
@@ -180,7 +180,7 @@ class ROMClassWriter::AnnotationWriter :
 
 	void visitAnnotation(U_16 typeIndex, U_16 elementValuePairCount)
 	{
-		_cursor->writeBigEndianU16(_constantPoolMap->getROMClassCPIndexForReference(typeIndex), Cursor::GENERIC);
+		_cursor->writeBigEndianU16(_constantPoolMap->getROMClassCPIndexForAnnotationWithCheck(typeIndex), Cursor::GENERIC);
```
Comment (DanHeidinga) [inline runtime/bcutil/ROMClassWriter.cpp]: Doesn't this still return 0 as the cpIndex for annotations with invalid type_index? I thought that was why the previous PR was closed:
`We realized that 0 is a valid constant pool entry. Closing this while we develop another solution.`

Comment (pdbain-ibm) [reply runtime/bcutil/ROMClassWriter.cpp]: The prior solution set the classfile constant pool index to zero (a valid entry).

This change set leaves the classfile constant pool index as-is but does not mark it as used by an annotation.  When we create the ROM constant pool, we check the validity of the  classfile index and force the ROM constant pool index to zero (i.e. not a valid index).   @fengxue-IS please would you please verify my logic?

Comment (fengxue-IS) [reply runtime/bcutil/ROMClassWriter.cpp]: The index zero for classfile CP is the first entry read, where as the zero index in ROM CP is a reserved null entry (which should trigger a runtime exception when attempting to read from it), in the previous PR, we set the classfile cp to zero, so this line of code with attempt to get an value the may be initialized (ie. classfile CP zero entry is not referenced)

File: runtime/bcutil/ConstantPoolMap.hpp [hunk 2]
```diff
@@ -210,6 +210,13 @@ class ConstantPoolMap
 
 	U_16 getROMClassCPIndexForReference(U_16 cfrCPIndex)  const { return getROMClassCPIndex(cfrCPIndex); }
 	U_16 getROMClassCPIndexForAnnotation(U_16 cfrCPIndex) const { return getROMClassCPIndex(cfrCPIndex); }
+	U_16 getROMClassCPIndexForAnnotationWithCheck(U_16 cfrCPIndex) const { 
+		if ((getCPTag(cfrCPIndex) == CFR_CONSTANT_Utf8) && isMarked(cfrCPIndex, ANNOTATION)) {
+			 return getROMClassCPIndex(cfrCPIndex); 
+		} else {
+			return 0;
+		}
+	}
```
Comment (gacholio) [inline runtime/bcutil/ConstantPoolMap.hpp]: Multiple returns are discouraged, especially with code this simple.  How about initializing a return value to 0, and setting it on success?

File: runtime/bcutil/ROMClassWriter.cpp (hunk 1 above)
Comment (DanHeidinga) [reply runtime/bcutil/ROMClassWriter.cpp]: The zero index in constantpool means null, even in a classfile.  See JVMS 4.1
> The `constant_pool` table is indexed from 1 to `constant_pool_count-1`

**Comment by DanHeidinga** [comment]
This is basically the same as #1579 but less clear.  Please reopen #1579 and we'll proceed with that item.

**Comment by DanHeidinga** [comment]
I suspect that's due to having closed the other pr and then creating a new one with the same branch name.

Here's a link to the patch for the original PR: https://patch-diff.githubusercontent.com/raw/eclipse/openj9/pull/1579.patch

File: runtime/bcutil/ROMClassWriter.cpp (hunk 1 above)
Comment (pdbain-ibm) [reply runtime/bcutil/ROMClassWriter.cpp]: There is no defined mapping from classfile offset 0, as it is never used, so we felt it was safer to leave a valid classfile index and detect the error when we do the mapping.

File: runtime/bcutil/ConstantPoolMap.hpp (hunk 2 above)
Comment (pdbain-ibm) [reply runtime/bcutil/ConstantPoolMap.hpp]: Done.  Thanks.

**Comment by pdbain-ibm** [comment]
> I suspect that's due to having closed the other pr and then creating a new one with the same branch name.
I created the new  pull request because I couldn't reopen the old one.

File: runtime/bcutil/ROMClassWriter.cpp (hunk 1 above)
Comment (DanHeidinga) [reply runtime/bcutil/ROMClassWriter.cpp]: index 0 is used as a null value.  We're safe to use it here.

**Comment by pdbain-ibm** [comment]
Reverted to the previous changeset.

File: runtime/bcutil/ROMClassWriter.cpp (hunk 1 above)
Comment (pdbain-ibm) [reply runtime/bcutil/ROMClassWriter.cpp]: In that case, I reverted to the previous changeset.

**Comment by DanHeidinga** [comment]
Peter and I had an at desk discussion about this PR and I now understand the cause better.  He'll update the description and add a comment to the code describing why we need to swizzle to cpIndex 0 rather than using the existing cpIndex.  Then this will be ready to merge

**Comment by DanHeidinga** [comment]
See also the comments in some of the marking code for method refs / field refs which indicates the same problem with a TODO
ie:
```
ClassFileOracle::markMethodRefAsReferenced(U_16 cpIndex)
{
	Trc_BCU_Assert_True((CFR_CONSTANT_Methodref == _classFile->constantPool[cpIndex].tag) || (CFR_CONSTANT_InterfaceMethodref == _classFile->constantPool[cpIndex].tag));

	// TODO markConstantAsReferenced(cpIndex); -- this breaks the iteration strategy in ROMClassBuilder
```

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8

**Pull request merged**